### ISSUE TITLE
from-290-fix-gating-on-Linformer-block

### ISSUE THREAD

**Issue by david-thrower** [david-thrower/cerebros-core-algorithm-alpha#291]
# On the composite LLM, fix Inconsistent Gating in the Linformer Block

## From #290

There is an inconsistency in how blocks merge their input and output streams. Most blocks use the GatedMergeLayer, but LinformerBlock uses a custom, less sophisticated implementation.

Problem: The LinformerBlock uses a simple sigmoid gate to blend the original input and the attention output. This is inconsistent with the other blocks which use the GatedMergeLayer. Consistency in architecture is important for stable gradient flow, especially when the layers are stacked.

Solution: Modify LinformerBlock to use the same GatedMergeLayer as the others.

In llm_utils.py, modify the LinformerBlock class:

# AI Suggested solution to review and integrate:

```
@tf.keras.utils.register_keras_serializable(package='cerebrosllmutils', name='LinformerBlock')
class LinformerBlock(tf.keras.layers.Layer):
    """
    A Transformer Block using Pre-Layer Normalization and the LinformerAttention layer.
    This version uses a GATING mechanism for stream merging.
    """

    def __init__(self, d_model, k_proj, dff, dropout_rate=0.1, ffn_dropout_rate=0.1, **kwargs):
        super().__init__(**kwargs)
        self.d_model = d_model
        self.k_proj = k_proj
        self.dff = dff
        self.dropout_rate = dropout_rate
        self.ffn_dropout_rate = ffn_dropout_rate

        # --- Attention Sub-layer ---
        self.attention = LinformerAttention(
            d_model=d_model,
            k_proj=k_proj,
            name="linformer_attention"
        )
        self.dropout1 = tf.keras.layers.Dropout(dropout_rate)
        self.layernorm1 = tf.keras.layers.LayerNormalization(epsilon=1e-6)

        # --- Stream Merging Layer (GATING) ---
        # *** CHANGE: Use the standard GatedMergeLayer for consistency ***
        self.gate = GatedMergeLayer(d_model)

        # --- Feed-Forward Network (FFN) Sub-layer ---
        self.ffn = tf.keras.Sequential([
            tf.keras.layers.Dense(dff, activation='relu'),  # (batch, seq_len, dff)
            tf.keras.layers.Dense(d_model)  # (batch, seq_len, d_model)
        ], name="feed_forward_network")
        self.dropout2 = tf.keras.layers.Dropout(ffn_dropout_rate)
        self.layernorm2 = tf.keras.layers.LayerNormalization(epsilon=1e-6)

    def call(self, inputs, training=False):
        # --- Attention Sub-layer with Pre-LN and Gated Stream Merging ---
        # 1. Normalize inputs
        norm_x = self.layernorm1(inputs)
        # 2. Apply attention
        attn_output = self.attention(norm_x)
        # 3. Apply dropout
        attn_output = self.dropout1(attn_output, training=training)

        # 4. *** CHANGE: GATE the original input and the attention output using the standard layer ***
        # This replaces the old manual gating logic.
        merged_output = self.gate([inputs, attn_output])

        # --- Feed-Forward Sub-layer with Pre-LN and Residual ---
        # 1. Normalize the output of the merged stream
        norm_merged = self.layernorm2(merged_output)
        # 2. Apply FFN
        ffn_output = self.ffn(norm_merged)
        # 3. Apply dropout
        ffn_output = self.dropout2(ffn_output, training=training)
        # 4. Residual connection
        final_output = merged_output + ffn_output

        return final_output

    def get_config(self):
        config = super().get_config()
        config.update({
            "d_model": self.d_model,
            "k_proj": self.k_proj,
            "dff": self.dff,
            "dropout_rate": self.dropout_rate,
            "ffn_dropout_rate": self.ffn_dropout_rate,
        })
        return config
```
